src/attr/_make.py

```diff
@@ -1103,7 +1099,10 @@ def _make_hash(cls, attrs, frozen, cache_hash):
     unique_filename = _generate_unique_filename(cls, "hash")
     type_hash = hash(unique_filename)
 
-    method_lines = ["def __hash__(self):"]
+    method_lines = [
+        "from attr._make import _CacheHashWrapper",
```

Comment:
I don't think I've ever seen an import inside a class definition before - this is a clever way to work around the fact that we need to access something from inside `attr` but this code will be `eval`'d in a namespace we don't own and can't control.  But, to avoid subclasses seeing this, it should probably be `del`'d after we're done defining `__hash__`.  An alternative would be:

```py
def __hash__(self, _cache_wrapper=__import__('attr._make')._make._CacheHashWrapper):
```

That's one line instead of 3, and `__import__` is maybe a bit less weird than an import inside a class body - at least, I've seen it more.

Reply:
Apparently this would normally give import errors because `_make_hash(` is actually called from `attr._make`, leading to a dependency cycle that apparently is resolved by using the `import` syntactic sugar.

However, that led me to the realization that the current code unconditionally imports `_CacheHashWrapper` and unconditionally wraps the hash value in `_cache_wrapper`, even for `cache_hash=False` classes.

Since the calls to `_make_hash` in `attr` don't use `cache_hash=True`, switching over to conditionally including those features (which should be done anyway) fixes that problem.